provider/oracle/bootstrap.go

```diff
@@ -0,0 +1,172 @@
+	"github.com/juju/juju/instance"
+)
+
+// launchBootstrapController creates a new vm inside
+// the oracle infrastracuture and parses the response into a instance.Instance
+func launchBootstrapConstroller(c *oci.Client, params []oci.InstanceParams) (instance.Instance, error) {
+	if c == nil {
+		return nil, errors.NotFoundf("client")
+	}
+
+	logger.Infof("Launching tbe bootstrap creation method")
+	resp, err := c.CreateInstance(params)
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+
+	fmt.Printf("%v\n", resp)
+	os.Exit(1)
+	instance, err := newInstance(resp)
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+
+	return instance, nil
+}
+
+// uploadSSHControllerKeys checks if the oracle api has the present key if not
+// then it will upload it, if there's arealdy the key we should check if the key
+// there maches the key we want to add, if not then we should modify the content
+// in order to be sure we are using the correct key
+func uploadSSHControllerKeys(c *oci.Client, key string) (string, error) {
+	if c == nil {
+		return "", errors.Errorf("cannot use nil error client to upload ssh keys")
+	}
+
+	if len(key) == 0 {
+		return "", errors.Errorf("cannot use empty key in order to upload it")
+	}
+
+	resp, err := c.SSHKeyDetails("juju-client-key")
+	if err != nil {
+		if strings.Contains(err.Error(), "does not exist") {
+			// the keys does not exist, we should upload them
+			resp, err = c.AddSHHKey("juju-client-key", key, true)
+			if err != nil {
+				return "", errors.Errorf("cannot upload the juju client keys")
+			}
+		} else {
+			return "", errors.Errorf("cannot get details of the juju-client-key")
+		}
+	}
+
+	keyrsponse := key[:strings.Index(key, "juju-client-key")-1]
+	if !resp.Enabled || resp.Key != keyrsponse {
+		if !resp.Enabled {
+			logger.Debugf("The Enabled flag is set to false on the key, setting it to true")
+		}
+		if resp.Key != keyrsponse {
+			logger.Debugf("The key provided with the oracle one does not match, update the key content")
+		}
+		resp, err = c.UpdateSSHKey("juju-client-key", key, true)
+		if err != nil {
+			return "", errors.Errorf("cannot enable of update the juju-client-key content")
+		}
+		if !resp.Enabled {
+			return "", errors.Errorf("juju client key is not enabled after previously tried to enable")
+		}
+	}
+
+	return resp.Name, nil
+}
+
+// isComplaint takes a ImageList object and finds
+// if it contains the given shapName, imageVersion and imageArch
+// if not it will return false
+func isComplaint(
+	list response.ImageList,
+	shapeName, imageVersion, imageArch string,
+) bool {
+
+	i := 0
+	for _, val := range list.Entries {
+		if strings.Contains(
+			val.Attributes.SupportedShapes,
+			shapeName,
+		) {
+			i++
+			break
+		}
+	}
+	if strings.Contains(list.Name, imageVersion) {
+		i++
+	}
+	if strings.Contains(list.Name, imageArch) {
+		i++
+	}
+	if i == 3 {
+		return true
+	}
+
+	return false
+}
+
+// checkImageList checks if there is a image specified in the
```

Comment:
I don't quite understand the point of this function. The image metadata should be complete and good enough to base a decision on.

We should expect that the image metadata is generated by querying the cloud's image API. So it doesn't make sense to query that API *again* to cross-check, since they should be identical. If they're not identical, that means the image metadata is out of date, and needs to be updated externally.